File: src/services/api/credentialService.ts
/**
 * Credential Service - API calls for credential management
 */

import apiClient from './apiClient';
import type { Credential, CredentialType } from '@/types';

interface CredentialApiResponse {
  id: string;
  type: string;
  issuer: string;
  subject: string;
  issued_at: string;
  expires_at: string;
  status: string;
  claims: Record<string, unknown>;
}

function transformCredentialResponse(response: CredentialApiResponse): Credential {
  return {
    id: response.id,
    type: response.type as CredentialType,
    issuer: response.issuer,
    subject: response.subject,
    issuedAt: response.issued_at,
    expiresAt: response.expires_at,
    status: response.status as Credential['status'],
    claims: response.claims,
  };
}

interface CredentialVerifyResponse {
  credentialId: string;
  isValid: boolean;
  status: Credential['status'];
  issuer: string;
  subject: string;
  type: CredentialType;
  expiresAt: string;
  reason: string | null;
}

export const credentialService = {
  /**
   * Issue a ShipownerVerificationCredential
   */
  issueShipownerCredential: async (data: {
    shipownerDid: string;
    companyName: string;
    registrationNumber: string;
    country: string;
    documentsVerified?: string[];
  }): Promise<Credential> => {
    const request = {
      shipowner_did: data.shipownerDid,
      company_name: data.companyName,
      registration_number: data.registrationNumber,
      country: data.country,
      documents_verified: data.documentsVerified || ['certificate_of_incorporation', 'registry_extract'],
    };
    const response = await apiClient.post<CredentialApiResponse>('/credentials/shipowner/verify', request);
    return transformCredentialResponse(response);
  },

  /**
   * Issue a VesselOwnershipCredential
   */
  issueVesselOwnershipCredential: async (data: {
    vesselId: string;
    ownerDid: string;
    ownershipType?: string;
  }): Promise<Credential> => {
    const request = {
      vessel_id: data.vesselId,
      owner_did: data.ownerDid,
      ownership_type: data.ownershipType || 'registered_owner',
    };
    const response = await apiClient.post<CredentialApiResponse>('/credentials/vessel/ownership', request);
    return transformCredentialResponse(response);
  },

  /**
   * Get a credential by ID
   */
  getCredential: async (credentialId: string): Promise<Credential> => {
    const response = await apiClient.get<CredentialApiResponse>(`/credentials/${credentialId}`);
    return transformCredentialResponse(response);
  },

  /**
   * Verify if a credential is valid
   */
  verifyCredential: async (credentialId: string): Promise<CredentialVerifyResponse> => {
    const response = await apiClient.get<{
      credential_id: string;
      is_valid: boolean;
      status: string;
      issuer: string;
      subject: string;
      type: string;
      expires_at: string;
      reason: string | null;
    }>(`/credentials/verify/${credentialId}`);
    return {
      credentialId: response.credential_id,
      isValid: response.is_valid,
      status: response.status as Credential['status'],
      issuer: response.issuer,
      subject: response.subject,
      type: response.type as CredentialType,
      expiresAt: response.expires_at,
      reason: response.reason,
    };
  },

  /**
   * Revoke a credential
   */
  revokeCredential: async (credentialId: string, reason: string): Promise<Credential> => {
    const response = await apiClient.post<CredentialApiResponse>('/credentials/revoke', {
      credential_id: credentialId,
      reason,
    });
    return transformCredentialResponse(response);
  },

  /**
   * Get all credentials for a subject (shipowner or vessel)
   */
  getCredentialsBySubject: async (subjectDid: string): Promise<Credential[]> => {
    const encodedDid = encodeURIComponent(subjectDid);
    const response = await apiClient.get<CredentialApiResponse[]>(`/credentials/subject/${encodedDid}`);
    return response.map(transformCredentialResponse);
  },

  /**
   * Get all credentials of a specific type
   */
  getCredentialsByType: async (credentialType: CredentialType): Promise<Credential[]> => {
    const response = await apiClient.get<CredentialApiResponse[]>(`/credentials/type/${credentialType}`);
    return response.map(transformCredentialResponse);
  },

  /**
   * List all credentials
   */
  listCredentials: async (): Promise<Credential[]> => {
    const response = await apiClient.get<CredentialApiResponse[]>('/credentials/');
    return response.map(transformCredentialResponse);
  },
};

export default credentialService;

File: backend/services/credential_service.py
"""
Credential service for issuing and managing verifiable credentials.
"""

from datetime import datetime, timedelta
from typing import Optional, List
import uuid

from models.credential import (
    Credential, CredentialType, CredentialStatus,
    ShipownerCredentialRequest, VesselOwnershipCredentialRequest,
    CredentialResponse
)
from services.json_store import JSONStore
from config import (
    CREDENTIALS_FILE, PLATFORM_DID, CREDENTIAL_VALIDITY_DAYS,
    SHIPOWNER_CREDENTIAL_TYPE, VESSEL_OWNERSHIP_CREDENTIAL_TYPE
)


class CredentialService:
    """Service for credential operations."""

    def __init__(self):
        self.store = JSONStore(CREDENTIALS_FILE)

    def issue_shipowner_credential(self, data: ShipownerCredentialRequest) -> CredentialResponse:
        """
        Issue a ShipownerVerificationCredential.
        This credential verifies that the platform has verified the shipowner's identity.
        """
        credential_id = f"CRED-{str(uuid.uuid4())[:8].upper()}"
        now = datetime.utcnow()
        expires = now + timedelta(days=CREDENTIAL_VALIDITY_DAYS)

        credential = {
            'id': credential_id,
            'type': CredentialType.SHIPOWNER_VERIFICATION.value,
            'issuer': PLATFORM_DID,
            'subject': data.shipowner_did,
            'issued_at': now.isoformat(),
            'expires_at': expires.isoformat(),
            'status': CredentialStatus.ACTIVE.value,
            'claims': {
                'company_name': data.company_name,
                'registration_number': data.registration_number,
                'country': data.country,
                'kyc_level': 'enhanced',
                'documents_verified': data.documents_verified
            },
            'revocation_reason': None,
            'revoked_at': None
        }

        self.store.create(credential)
        return CredentialResponse(**credential)

    def issue_vessel_ownership_credential(
        self,
        data: VesselOwnershipCredentialRequest,
        vessel_did: str,
        vessel_imo: str,
        vessel_name: str,
        owner_company_name: str,
        verified_documents: List[str]
    ) -> CredentialResponse:
        """
        Issue a VesselOwnershipCredential.
        This credential attests that a shipowner owns a specific vessel.
        """
        credential_id = f"CRED-{str(uuid.uuid4())[:8].upper()}"
        now = datetime.utcnow()
        expires = now + timedelta(days=CREDENTIAL_VALIDITY_DAYS)

        credential = {
            'id': credential_id,
            'type': CredentialType.VESSEL_OWNERSHIP.value,
            'issuer': PLATFORM_DID,
            'subject': data.owner_did,
            'issued_at': now.isoformat(),
            'expires_at': expires.isoformat(),
            'status': CredentialStatus.ACTIVE.value,
            'claims': {
                'vessel_did': vessel_did,
                'vessel_imo': vessel_imo,
                'vessel_name': vessel_name,
                'owner_did': data.owner_did,
                'owner_company_name': owner_company_name,
                'ownership_type': data.ownership_type,
                'verified_documents': verified_documents
            },
            'revocation_reason': None,
            'revoked_at': None
        }

        self.store.create(credential)
        return CredentialResponse(**credential)

    def get_credential(self, credential_id: str) -> Optional[CredentialResponse]:
        """Get a credential by ID."""
        record = self.store.get_by_id(credential_id)
        if record:
            return CredentialResponse(**record)
        return None

    def verify_credential(self, credential_id: str) -> dict:
        """
        Verify if a credential is valid.
        Checks: exists, not revoked, not expired.
        """
        record = self.store.get_by_id(credential_id)

        if not record:
            return {
                'credential_id': credential_id,
                'is_valid': False,
                'status': None,
                'reason': 'Credential not found'
            }

        # Check if revoked
        if record['status'] == CredentialStatus.REVOKED.value:
            return {
                'credential_id': credential_id,
                'is_valid': False,
                'status': CredentialStatus.REVOKED.value,
                'issuer': record['issuer'],
                'subject': record['subject'],
                'type': record['type'],
                'expires_at': record['expires_at'],
                'reason': f"Credential revoked: {record.get('revocation_reason', 'No reason provided')}"
            }

        # Check if expired
        expires_at = datetime.fromisoformat(record['expires_at'].replace('Z', '+00:00'))
        if datetime.utcnow() > expires_at.replace(tzinfo=None):
            # Update status to expired
            self.store.update(credential_id, {'status': CredentialStatus.EXPIRED.value})
            return {
                'credential_id': credential_id,
                'is_valid': False,
                'status': CredentialStatus.EXPIRED.value,
                'issuer': record['issuer'],
                'subject': record['subject'],
                'type': record['type'],
                'expires_at': record['expires_at'],
                'reason': 'Credential has expired'
            }

        # Credential is valid
        return {
            'credential_id': credential_id,
            'is_valid': True,
            'status': CredentialStatus.ACTIVE.value,
            'issuer': record['issuer'],
            'subject': record['subject'],
            'type': record['type'],
            'expires_at': record['expires_at'],
            'reason': None
        }

    def revoke_credential(self, credential_id: str, reason: str) -> Optional[CredentialResponse]:
        """Revoke a credential."""
        record = self.store.get_by_id(credential_id)
        if not record:
            return None

        updated = self.store.update(credential_id, {
            'status': CredentialStatus.REVOKED.value,
            'revocation_reason': reason,
            'revoked_at': datetime.utcnow().isoformat()
        })

        if updated:
            return CredentialResponse(**updated)
        return None

    def list_credentials_by_subject(self, subject_did: str) -> List[CredentialResponse]:
        """List all credentials for a subject (shipowner or vessel)."""
        records = self.store.find_by_field('subject', subject_did)
        return [CredentialResponse(**r) for r in records]

    def list_credentials_by_type(self, credential_type: CredentialType) -> List[CredentialResponse]:
        """List all credentials of a specific type."""
        records = self.store.find_by_field('type', credential_type.value)
        return [CredentialResponse(**r) for r in records]

    def list_all_credentials(self) -> List[CredentialResponse]:
        """List all credentials."""
        records = self.store.get_all()
        return [CredentialResponse(**r) for r in records]
